### PR TITLE
Best Case time complexity for Bubble Sort

### PR DESCRIPTION
**Pull request by ashu01**
*  Best case time complexity is Ω(n). 
*  Space Optimized fibonacci series
*  Also added this to unit test

### CODE REVIEW

File: pygorithm/fibonacci/space_optimized.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        for i in range(1, n):
+            c = a + b
+            a = b
+            b = c
+        return b
+
+    def sequence(n):
+        """
+        Return sequence if Fibonacci values as list.
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: Not `if`, but `of`. I also missed it my code (i will change it in next `PR`).

File: pygorithm/fibonacci/space_optimized.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        if n == 0:
+            return a
+
+        if n == 1:
+            return b
+
+        for i in range(1, n):
+            c = a + b
+            a = b
+            b = c
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: Clean up it like that:

```python
a, b = b, a + b
```

File: pygorithm/fibonacci/space_optimized.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return b
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: Make empty space around `return` statement in this case, please.

File: pygorithm/fibonacci/space_optimized.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        a = 0
+        b = 1
+
+        if n == 0:
+            return a
+
+        if n == 1:
+            return b
+
+        for i in range(1, n):
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: You actually do not need `if` statements and `1` point at fist argument to `range`.

Make it simple like:
```python
    a, b= 0, 1

    for _ in range(n):
        a, b = b, a + b
    return a
```

**Comment by ashu01** [comment]
done.

File: pygorithm/fibonacci/space_optimized.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Author: Ashutosh Gupta
+# Created On: 8/15/2017
+# Time: 3:24 PM
+
+"""
+nth fibonacci number - Space Optimized
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: Can we call it `Fibonacci implementation through using looping technique (space optimized)` or something else?

File: pygorithm/fibonacci/space_optimized.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+"""
+
+import inspect
+
+
+def get_sequence(n):
+    """
+    Return Fibonacci sequence from zero to specified number as list.
+    """
+
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: Go ahead and follow one style — no space after function (or method in class) documentation.

File: pygorithm/fibonacci/space_optimized.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+def get_code():
+    """
+    Return source code of Fibonacci sequence logic's implementation.
+    """
+    return inspect.getsource(get_sequence)
+
+# Time Complexity: O(n)
+# Extra Space: O(1)
```
Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: Can we move complexity information to better place, for example to module documentation (as third line).
I think it is important thing.

Comment (ashu01) [inline pygorithm/fibonacci/space_optimized.py]: Its basically space optimized technique.
in cache.py, Extra space is O(n) and same in  recursive method.
The best suited name is space_optimized as far as i think.

Comment (ashu01) [inline pygorithm/fibonacci/space_optimized.py]: Where to put the complexity information ?

Comment (ashu01) [inline pygorithm/fibonacci/space_optimized.py]: can you suggest ?

Comment (dmytrostriletskyi) [inline pygorithm/fibonacci/space_optimized.py]: > Where to put the complexity information ?

Before:
```
"""
Fibonacci implementation through ***.
"""
```

After:
```
"""
Fibonacci implementation through ***.

Time Complexity: O(n). Extra Space: O(1)
"""
```

I think it seems great.

**Comment by OmkarPathak** [comment]
@ashu01 this PR is having many conflicts. Can you please make a new PR only with the file `pygorithm/factorial/space_optimised.py`. Then it would be easy to merge them

**Pull request closed without merging**